### ISSUE TITLE
Improved Domain and IP Lookups

### ISSUE THREAD

**Issue by pmisner** [brimdata/brim#1714]
I'm always frustrated when I right click on the nslookup tool in Brim. I know it's Microsoft's and I could configure it to get better results, but now mostly I see MarkMonitor blocks. 

Right now, I use a suite of tools at [https://www.nirsoft.net/domain_lookup_tools.html] to augment the information I get in Brim. These tools provide IP and DNS Information from multiple sources.  Some of the tools allow for bulk entry of IPs. That would be really cool if you could extract a block of external IPs and search them in bulk.

The code is open source and free to use, but I know you'll give Mr. Nir a couple of shares if you integrate it into the project.

Thank you once again for a great tool.

**Comment by philrz** [comment]
@pmisner: Thanks for the ideas. I'll check them out. One point of clarification: You spoke of a right-click "nslookup" option, but I don't believe we've ever had that. Did you mean Whois?

**Comment by philrz** [comment]
@pmisner: Thanks for bringing this to our attention. Indeed, the initial right-click "Whois" support was created as a quick proof-of-concept and in many ways was put there to inspire ideas. These tools provide some nice food for thought along those lines. I did some hacking with them just now, and here's some thoughts on directions we might take with them.

From the [link](https://www.nirsoft.net/domain_lookup_tools.html) you provided, I downloaded the [ZIP package](https://www.nirsoft.net/packages/domainlookuptools.zip) linked from the top of the page and started by playing with [DomainHostingView](https://www.nirsoft.net/packages/domainlookuptools.zip) since it happened to be the first one listed. In thinking about it as something that could be invoked as a right-click option, I went looking for either an API or a way to invoke it on the command line, and thankfully the latter was easy to find.

Following the example shown there, I was able to do:

```
C:\> DomainHostingView.exe /GenerateReport brimsecurity.com c:\temp\output.html
```

And as promised, this produced the HTML (which I've attached in full as compressed [output.html.gz](https://github.com/brimdata/brim/files/6698070/output.html.gz)) which renders fine in a browser. Some pop-ups in Brim already contain renderer HTML (such as the Release Notes tab described in #1594 that'll appear in the next GA release) so I expect it would be trivial for us to render this output in the Brim app.

![image](https://user-images.githubusercontent.com/5934157/123016443-75776b80-d37f-11eb-817b-b23e8f623d42.png)

Rather than hard-coding something specific like we did with Whois, I'd be tempted to promote some kind of flexible configuration the user could invoke. This is just a straw man proposal, but perhaps there could be config in **Preferences** that allows the user to specify all the current "callouts" available via-right-click. Each entry could include:

* The name of the "callout" as it would appear in the right-click menu, e.g., "NirSoft Domain Hosting View"
* A list of Zed data types of field values on which the command would be made available
* The command line, with some token that would be populated with the value on which the user right-clicked, e.g.:
      ```
      C:\NirSoft\DomainHostingView.exe /GenerateReport $RIGHT_CLICKED_VALUE c:\temp\output.html
      ```
* To cover this specific tool, it seems we might benefit from having some other token like `$TMPFILE` to use in place of the hard-coded `c:\temp\output.html` shown above. That way it could be populated by Brim with a path to a temporary file it asks the OS to create, so that way the tool could send the output somewhere unique and then Brim could find the output there and delete the temp file after the output has been shown to the user

In addition to creating this as config the user could see/modify by hand, we could conceivably also add some kind of "import" functionality so that a curated set of such configs could be loaded in bulk. This way users could (for example) get config for all the NirSoft tools at once. No doubt we could also ship Brim with that config already loaded, but as we want this all to be community-friendly, I'm looking for ways we can empower users to enhance these things on their own rather than having to wait on the core Dev team at Brim to add them as if they had to be monolithic, baked-in features.

Being able to do that for single values seems like it would be a first step. It seems a second-order feature request is also implied by the desire to "extract a block [...] and search them in bulk". For instance, I saw this tool from the set takes multiple domains:

```
C:\> DNSDataView.exe /Domains brimsecurity.com,corelight.com /shtml C:\temp\dns.html
```

![image](https://user-images.githubusercontent.com/5934157/123018059-b9b83b00-d382-11eb-8f62-cfb6085c4b2e.png)

Right now in Brim right-click is only tied to single values. I think we've talked before about ways to offer multi-select such as via Shift-Click/Ctrl-Click (or Command-Click on macOS), in which case perhaps we could cover this case by comma-separating all the values that had been clicked on (and depending on what we find in a more thorough audit of tools, maybe there'd need to be ways to vary how they're formatted, e.g. some other tool out there might want them separated by semicolons, spaces, etc.)

While it seems the architecture would be ready to support this, it does seem like a somewhat ambitious undertaking. But these do seem like useful tools and it would be validation of an approach that could hopefully be re-used i other contexts. I'll be sharing this with the core Dev team to see what they think.

**Comment by pmisner** [comment]
@philrz  I didn't think about making it configurable through preferences. I like the idea. There's probably several tools like this that can be hooked in. I realize that the multiples would be a problem.  Some type of way to pipe or export external IPs or domains  to a csv (with returns (columns)), or just in a big row might work for this. It could be a function generated after a user sets up the filter they want.

**Comment by philrz** [comment]
@pmisner recently suggested a variation on this in another [Slack thread](https://brimsec.slack.com/archives/C010DR0HHMF/p1625161746068700), so I'll capture it here and we can always break it out into a separate issue if that's justified when we sit down to implement. In their own words:

> In Excel, I use the hyperlink command to send IPs (as cell entries) to different OSINT solutions. For example "=HYPERLINK("https://viz.greynoise.io/query/?gnql="&A2,"Greynoise")" sends the IP in cell a2 to greynoise for a look.  I think it would be really cool to send an external IP as a variable similar to this.

In other words, whereas the original idea captured in this issue was configurable callouts to _programs_ that could act on a field value (e.g. Brim's current right-click "Whois" option is a hard-coded example of this), this would be a configurable callout to a _web_ _service_ that could act on a field value (which Brim's current right-click "VirusTotal" option is a hard-coded example of).

Ultimately both features would feel much the same: A right-click Whois option _could_ be implemented in either approach and the user could be blissfully unaware of which returns the info. But being able to support both approaches would allow for a wider set of possible callout functions.

**Comment by jameskerr** [comment]
These are all great ideas in my opinion. I'd like to work on them!
